### PR TITLE
[SPARK-43259][SQL] Assign a name to the error class _LEGACY_ERROR_TEMP_2024

### PR DESCRIPTION
**Pull request by mihailomilosevic2001**
### What changes were proposed in this pull request?
In the PR, I propose to assign the proper name `INVALID_EXPRESSION_ENCODER`  to the legacy error class `_LEGACY_ERROR_TEMP_2024`, and add a test to the suite which uses `checkError()`. Also this PR improves the error message.

### Why are the changes needed?
Proper name improves user experience w/ Spark SQL.

### Does this PR introduce _any_ user-facing change?
Yes, the PR changes an user-facing error message.

### How was this patch tested?
By running the modified test suite:

```
./build/mvn -Dtest=none -DwildcardSuites=org.apache.spark.sql.errors.QueryExecutionErrorsSuite test
```

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala [hunk 1]
```diff
@@ -1151,6 +1153,21 @@ class QueryExecutionErrorsSuite
...
     )
   }
+
+  test("SPARK-43259: Uses unsupported custom encoder") {
+    class CustomEncoder extends Encoder[Int] {
+      override def schema: StructType = StructType(Array.empty[StructField])
+      override def clsTag: ClassTag[Int] = ClassTag.Int
+    }
+    val e = intercept[SparkRuntimeException] {
+      encoderFor(new CustomEncoder)
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala:1163]: Can you try to reproduce the issue using public Spark SQL API (sql or Java/Scala/Python api)? If it is not possible, we should convert the error to an internal one.

Comment (mihailom-db) [reply sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala]: I explored the problem and have changed the PR description accordingly, could you take a look at it again, if it makes sense to you as well? @MaxGekk

Comment (MaxGekk) [reply sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala]: I think this is an user facing error, see how to reproduce it using public API:
```scala
    implicit val kryoEncoder = new Encoder[KryoData] {
      override def schema: StructType = StructType(Array.empty[StructField])
      override def clsTag: ClassTag[KryoData] = ???
    }
    val ds = Seq(KryoData(1), KryoData(2)).toDS()
    ds.groupByKey(p => p).count().collect()
```
it fails with:
```java
Only expression encoders are supported for now.
org.apache.spark.SparkRuntimeException: Only expression encoders are supported for now.
	at org.apache.spark.sql.errors.QueryExecutionErrors$.unsupportedEncoderError(QueryExecutionErrors.scala:463)
	at org.apache.spark.sql.catalyst.encoders.package$.encoderFor(package.scala:34)
```

**Review by MaxGekk — CHANGES_REQUESTED** [review]
Please, assign a proper name for the error class, and add a test using public API.

File: sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala (hunk 1 above)
Comment (mihailom-db) [reply sql/core/src/test/scala/org/apache/spark/sql/errors/QueryExecutionErrorsSuite.scala]: I have introduced this change. Just out of curiosity, should we change the error format to pass the type of encoder that was used and then include it in the message?

**Comment by MaxGekk** [comment]
> In the PR, I propose to assign the proper name UNSUPPORTED_ERROR ...
 
@mihailom-db Did you really assign this name?

**Comment by MaxGekk** [comment]
@mihailom-db BTW, do you have an account at OSS JIRA? If so, could you leave a comment in https://issues.apache.org/jira/browse/SPARK-43259 - telling that you are working on this.

**Comment by mihailom-db** [comment]
Requested the access for the OSS JIRA account, should be able to comment soon.

File: common/utils/src/main/resources/error/error-classes.json [hunk 2]
```diff
@@ -3748,6 +3748,12 @@
     },
     "sqlState" : "0A000"
   },
+  "UNSUPPORTED_ENCODER" : {
+    "message" : [
+      "Found unsupported encoder. Try switching to expression encoder."
```
Comment (MaxGekk) [inline common/utils/src/main/resources/error/error-classes.json:3753]: nit:
```suggestion
      "Found unsupported encoder. Try switching to an expression encoder."
```

File: common/utils/src/main/resources/error/error-classes.json [hunk 3]
```diff
@@ -3748,6 +3748,12 @@
...
+    ],
+    "sqlState" : "0A000"
```
Comment (MaxGekk) [inline common/utils/src/main/resources/error/error-classes.json:3755]: The category `0A` belongs to:
```
    "0A": "feature not supported",
```
see `error-categories.json`. Any reasons to choose this one? How about `46` or `42`?

File: common/utils/src/main/resources/error/error-classes.json (hunk 2 above)
Comment (MaxGekk) [reply common/utils/src/main/resources/error/error-classes.json:3753]: I think we should help users and point out how to switch to an expression encoder. At least, let's point out the doc:
https://spark.apache.org/docs/latest/api/java/index.html?org/apache/spark/sql/Encoder.html

File: common/utils/src/main/resources/error/error-classes.json (hunk 3 above)
Comment (mihailom-db) [reply common/utils/src/main/resources/error/error-classes.json]: I have changed the category to `42` as I believe the error is possible to happen only if a user defines implicit `Encoder` for which an implicit `Encoder` does not already exist. I experimented with trying to use types `Int` or `String`, but realised that we already have implicit values defined and in those cases `.toDS()` call was marked as unresolved and the code did not even compile. `KryoData` seems to not have an implicit Encoder defined, so the added test is possible to happen. So I am not sure if I should change error name and/or message accordingly?

File: common/utils/src/main/resources/error/error-classes.json (hunk 2 above)
Comment (mihailom-db) [reply common/utils/src/main/resources/error/error-classes.json]: Done.

File: common/utils/src/main/resources/error/error-classes.json [hunk 4]
```diff
@@ -3748,6 +3748,12 @@
     },
     "sqlState" : "0A000"
   },
+  "UNSUPPORTED_ENCODER" : {
+    "message" : [
+      "Found unsupported encoder. Try switching to an expression encoder. (For more information on Encoders, consult: https://spark.apache.org/docs/latest/api/java/index.html?org/apache/spark/sql/Encoder.html)"
```
Comment (MaxGekk) [inline common/utils/src/main/resources/error/error-classes.json:3753]: Could you follow other items in the file for doc refs, see
https://github.com/apache/spark/blob/0d938de112dfe142f40d4c6f86cfa1e6e32210ec/common/utils/src/main/resources/error/error-classes.json#L1523

**Review by MaxGekk — CHANGES_REQUESTED** [review]

File: common/utils/src/main/resources/error/error-classes.json (hunk 4 above)
Comment (mihailom-db) [reply common/utils/src/main/resources/error/error-classes.json]: Generally, I understood that these are generated files that are being referenced, but I referenced the file with all error state support and then I introduced a link to Encoder doc there.

File: common/utils/src/main/resources/error/error-classes.json [hunk 5]
```diff
@@ -3748,6 +3748,12 @@
     },
     "sqlState" : "0A000"
   },
+  "UNSUPPORTED_ENCODER" : {
+    "message" : [
+      "Found unsupported encoder. Try switching to an expression encoder. (For more information consult: '<docroot>/sql-error-conditions.html')"
+    ],
+    "sqlState" : "42000"
```
Comment (MaxGekk) [inline common/utils/src/main/resources/error/error-classes.json:3755]: Let's assign some unused id like `42001`.

File: common/utils/src/main/resources/error/error-classes.json [hunk 6]
```diff
@@ -3748,6 +3748,12 @@
     },
     "sqlState" : "0A000"
   },
+  "UNSUPPORTED_ENCODER" : {
```
Comment (MaxGekk) [inline common/utils/src/main/resources/error/error-classes.json:3751]: I think we are dealing with invalid encoder not some unsupported one. Could you rename it to: `INVALID_EXPRESSION_ENCODER`

File: common/utils/src/main/resources/error/error-classes.json [hunk 7]
```diff
@@ -3748,6 +3748,12 @@
...
+    "message" : [
+      "Found unsupported encoder. Try switching to an expression encoder. (For more information consult: '<docroot>/sql-error-conditions.html')"
```
Comment (MaxGekk) [inline common/utils/src/main/resources/error/error-classes.json:3753]:
```suggestion
      "Found an invalid expression encoder. Expects an instance of `ExpressionEncoder` but got <encoderType>. For more information consult: '<docroot>/api/java/index.html?org/apache/spark/sql/Encoder.html'"
```

File: common/utils/src/main/resources/error/error-states.json [hunk 8]
```diff
@@ -2933,6 +2933,12 @@
         "standard": "Y",
         "usedBy": ["SQL/Foundation", "PostgreSQL", "Redshift", "Oracle", "SQL Server"]
     },
+    "42001": {
+        "description": "Invalid encoder error",
+        "origin": "SQL/Foundation",
+        "standard": "Y",
+        "usedBy": ["SQL/Foundation"]
+    },
```
Comment (mihailom-db) [inline common/utils/src/main/resources/error/error-states.json:2941]: @MaxGekk I am not sure how I should approach to naming new sqlstate, do you have some more info on what these fields are about?

File: common/utils/src/main/resources/error/error-states.json [hunk 9]
```diff
@@ -2933,6 +2933,12 @@
         "standard": "Y",
         "usedBy": ["SQL/Foundation", "PostgreSQL", "Redshift", "Oracle", "SQL Server"]
     },
+    "42001": {
+        "description": "Invalid encoder error",
+        "origin": "SQL/Foundation",
+        "standard": "Y",
+        "usedBy": ["SQL/Foundation"]
```
Comment (MaxGekk) [inline common/utils/src/main/resources/error/error-states.json:2940]: Could you check is this SQL state used by other DBMS?

Comment (mihailom-db) [reply common/utils/src/main/resources/error/error-states.json]: I took a look at different DBMS sqlstates and to my knowledge it seems that this one is not used by any. Changed doc accordingly.

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @mihailom-db.

**Comment by MaxGekk** [comment]
@mihailom-db Congratulations with your first contribution to Apache Spark!

**Review by dongjoon-hyun — COMMENTED** [review]
Hi, @mihailom-db and @MaxGekk 

Does this PR pass the CIs? It seem to break Apache Spark `master` branch. Could you take a look at the failure?

**Comment by dongjoon-hyun** [comment]
```
[info] SparkThrowableSuite:
[info] - No duplicate error classes (20 milliseconds)
[info] - Error classes are correctly formatted (35 milliseconds)
[info] - SQLSTATE is mandatory (2 milliseconds)
[info] - Error category and error state / SQLSTATE invariants (23 milliseconds)
[info] - Message invariants (7 milliseconds)
[info] - Message format invariants (8 milliseconds)
[info] - Error classes match with document *** FAILED *** (55 milliseconds)
[info]   "...ects an instance of [ExpressionEncoder] but got `<encoderTy..." did not equal "...ects an instance of [`ExpressionEncoder`] but got `<encoderTy..." The error class document is not up to date. Please regenerate it. (SparkThrowableSuite.scala:353)
[info]   Analysis:
[info]   "...ects an instance of [ExpressionEncoder] but got `<encoderTy..." -> "...ects an instance of [`ExpressionEncoder`] but got `<encoderTy..."
```

**Comment by dongjoon-hyun** [comment]
NVM. Let me make a quick followup.

**Comment by dongjoon-hyun** [comment]
Here is a follow-up.
- #45226

**Pull request closed without merging**